contributor-docs/adrs/adr-004-children-as-api.md

```diff
@@ -0,0 +1,403 @@
+
+In the above case, if we want to make the recommended path easier and other paths harder, we can change the API to look something like this -
+
+```jsx
+<Flash variant="success" icon={CheckIcon}>
+  Changes saved!
+</Flash>
+```
+
+We are still using `children` for text content, but we have moved the `icon` back as a prop with reduced flexibility.
+
+When intentionally going off the happy path, developers can still drop down an abstraction level add an `Icon` to `children`, they would have to pick up the additional effort of setting compatible color, size and margin themselves. However, when it’s unintentional, it would feel like way too much work that the component should be doing automatically.
+
+```jsx
+<Flash variant="success">
+  <CheckIcon size={20} color="success.emphasis" sx={{marginRight: 2}} />
+  Changes saved!
+</Flash>
+```
+
+<br/>
+
+_Sidenote: We might want to name this prop `leadingIcon`, even if there is no `trailingIcon`. Consistent names across components plays a big role in making the API predictable._
+
+
+
+---
+
+<br/>
+
+You can see this pattern used in `NewButton`:
+
+The icon gets its color and margin based on the variant and size of the `Button`. This is the happy path we want folks to be on, so we ask for the icon component instead of asking the developer to render the icon.
+
+<img width="141" alt="image 9" src="https://user-images.githubusercontent.com/1863771/144945219-a853ed1c-f21d-412e-a388-6d74ec436645.png">
+
+```jsx
+<NewButton leadingIcon={SearchIcon}>Search</NewButton>
+<NewButton leadingIcon={SearchIcon} variant="primary" size="large">Search</NewButton>
+```
+
+```jsx
+// we prefer this:
+<NewButton leadingIcon={SearchIcon}>Search</NewButton>
+// over these:
+<NewButton><SearchIcon/> Search</NewButton>
+<NewButton leadingIcon={<SearchIcon/>}>Search</NewButton>
+```
+
+<br/>
+
+---
+
+<br/>
+
+## Flexibility is a spectrum and the case for composite components
+
+<br/>
+
+### 1. Scenarios where we want to restrict flexibility and bake-in design decisions for the most part, but allow some configuration.
+
+Consider this fake example:
+
+<img width="388" alt="image 7" src="https://user-images.githubusercontent.com/1863771/144945217-fd5eeab0-bfa5-4c53-9add-621dcd87b48c.png">
+
+We want users to be able to customise if the `Icon` is outline or filled. (I know I know the example is bit silly, but please go with it)
+
+<br/>
+
+Extending our `strict` API, we could add another prop to the component:
+
+```jsx
+<Flash variant="success" icon={CheckIcon} iconVariant="filled">
+  Changes saved!
+</Flash>
+```
+
+When we have an “element” and “elementProp” as props on a component, it’s a sign that we should create a child component here that is tied to the parent component:
+
+```jsx
+<Flash variant="success">
+  <Flash.Icon icon={CheckIcon} variant="filled" />
+  Changes saved!
+</Flash>
+```
+
+<br/>
+
+The `Parent.Child` syntax signals that this component is tied to the parent.
```

Comment:
Do we have composite components with nesting deeper than 2 levels? It might be good to have a recommendation for naming deeply nested children.

For example:
```
<ChoiceFieldset>
  <ChoiceFieldset.List>
    <ChoiceFieldset.Item value="labelOne">Label one</ChoiceFieldset.Item>
    <ChoiceFieldset.Item value="labelTwo">Label two</ChoiceFieldset.Item>
  </ChoiceFieldset.List>
</ChoiceFieldset
```

Reply:
we do! 

```jsx
<ActionList>
  <ActionList.Group>
      <ActionList.Item>
        <ActionList.LeadingVisual />
      </ActionList.Item>
  </ActionList.Group>
</ActionList>
```

the recommendation would be to attach components on the root level component and not middle ones, right?
`ActionList.LeadingVisual` and not `ActionList.Item.LeadingVisual`